File: non_primitives/non_primitives.py
"""
NON-PRIMITIVE DATA TYPES IN PYTHON
==================================

Non-primitive types are collections of items.
Main types:
1. List (Mutable, Ordered)
2. Tuple (Immutable, Ordered)
3. Set (Mutable, Unordered, Unique)
4. Dictionary (Mutable, Key-Value pairs)
"""

# ==========================================
# 1. Lists (Similar to JS Arrays)
# ==========================================
print("--- 1. Lists ---")
# Creation
numbers = [1, 2, 3, 4, 5]
mixed = [1, "two", {"id": 3}]

# Accessing
print(numbers[0]) # 1
print(len(numbers)) # 5

# Modifying
numbers.append(6) # Add to end
numbers.pop()     # Remove from end
numbers[0] = 100  # Modify element

# Looping Lists
print("Looping List:")
for num in numbers:
    print(num)

# Standard Loop (with index)
print("Looping with Index:")
for i in range(len(numbers)):
    print(f"Index {i}: {numbers[i]}")

# ==========================================
# 2. Tuples (Immutable Lists)
# ==========================================
print("\n--- 2. Tuples ---")
# Creation
coordinates = (10, 20)
single_item = (1,) # Note the comma

# Accessing
print(coordinates[0]) # 10

# Immutable
# coordinates[0] = 5 # TypeError: 'tuple' object does not support item assignment

# Use case: Return multiple values from function, keys in dictionary.

# ==========================================
# 3. Sets (Unique Collection)
# ==========================================
print("\n--- 3. Sets ---")
# Creation
unique_nums = {1, 2, 3, 3, 2, 1}
print(unique_nums) # {1, 2, 3} (Duplicates removed)

# Adding/Removing
unique_nums.add(4)
unique_nums.remove(1)

# Set Operations
a = {1, 2, 3}
b = {3, 4, 5}
print(a.union(b)) # {1, 2, 3, 4, 5}
print(a.intersection(b)) # {3}

# ==========================================
# 4. Dictionaries (Similar to JS Objects/Maps)
# ==========================================
print("\n--- 4. Dictionaries ---")
# Creation
user = {
    "name": "John",
    "age": 30,
    "is_admin": True
}

# Accessing
print(user["name"]) # "John"
print(user.get("city", "Unknown")) # Safe access with default

# Modifying
user["age"] = 31
user["city"] = "New York"
del user["is_admin"]

# Looping Dictionaries
print("Looping Dictionary:")
for key, value in user.items():
    print(f"{key}: {value}")

# ==========================================
# 5. List Comprehensions (Pythonic Map/Filter)
# ==========================================
print("\n--- 5. List Comprehensions (Map/Filter equivalent) ---")

items = [
    {"name": 'Bike', "price": 100},
    {"name": 'TV', "price": 200},
    {"name": 'Album', "price": 10},
    {"name": 'Book', "price": 5},
    {"name": 'Phone', "price": 500},
    {"name": 'Computer', "price": 1000},
]

# 1. FILTER equivalent
# Get items cheaper than 100
cheap_items = [item for item in items if item["price"] < 100]
print(f"Filtered (Cheap Items): {cheap_items}")

# 2. MAP equivalent
# Get just the names of items
item_names = [item["name"] for item in items]
print(f"Mapped (Names): {item_names}")

# 3. FIND equivalent (using next())
# Find the item with name 'Book'
# next(iterator, default)
found_item = next((item for item in items if item["name"] == 'Book'), None)
print(f"Found Item 'Book': {found_item}")

# 4. REDUCE equivalent
# Calculate total price of all items
# Needs import from functools
from functools import reduce
total = reduce(lambda current_total, item: item["price"] + current_total, items, 0)
print(f"Total Price (Reduce): {total}")

# Pythonic Sum (Simpler for this specific case)
total_sum = sum(item["price"] for item in items)
print(f"Total Price (Sum): {total_sum}")

File: non_primitives/non_primitives.js
/**
 * NON-PRIMITIVE DATA TYPES IN JAVASCRIPT
 * ======================================
 * 
 * Non-primitive types are objects. They are mutable and stored by reference.
 * Main types:
 * 1. Object
 * 2. Array
 */

// ==========================================
// 1. Objects
// ==========================================
console.log("--- 1. Objects ---");
// Creation
const user = {
    name: "John",
    age: 30,
    isAdmin: true,
    // Methods
    greet: function () {
        console.log(`Hello, I am ${this.name}`);
    }
};

// Accessing
console.log(user.name); // Dot notation
console.log(user["age"]); // Bracket notation

// Modifying
user.age = 31;
user.city = "New York"; // Adding new property
delete user.isAdmin;    // Deleting property

// Looping over Objects
console.log("Looping Object:");
for (let key in user) {
    console.log(`${key}: ${user[key]}`);
}

// ==========================================
// 2. Arrays
// ==========================================
console.log("\n--- 2. Arrays ---");
// Creation
const numbers = [1, 2, 3, 4, 5];
const mixed = [1, "two", { id: 3 }];

// Accessing
console.log(numbers[0]); // 1
console.log(numbers.length); // 5

// Looping Arrays
console.log("Looping Array (for...of):");
for (const num of numbers) {
    console.log(num);
}

// Standard loops
console.log("Looping Array (Standard for):");
for (let i = 0; i < numbers.length; i++) {
    console.log(numbers[i]);
}

// ==========================================
// 3. Array High-Order Methods
// ==========================================
console.log("\n--- 3. Array Methods (Functional) ---");

const items = [
    { name: 'Bike', price: 100 },
    { name: 'TV', price: 200 },
    { name: 'Album', price: 10 },
    { name: 'Book', price: 5 },
    { name: 'Phone', price: 500 },
    { name: 'Computer', price: 1000 },
];

console.log("Original Items:", items);

// 1. FILTER: Returns a new array with all elements that pass the test
// Get items cheaper than 100
const cheapItems = items.filter((item) => {
    return item.price < 100;
});
console.log("Filtered (Cheap Items):", cheapItems);

// 2. MAP: Returns a new array by applying a function to every element
// Get just the names of items
const itemNames = items.map((item) => {
    return item.name;
});
console.log("Mapped (Names):", itemNames);

// 3. FIND: Returns the value of the FIRST element that passes the test
// Find the item with name 'Book'
const foundItem = items.find((item) => {
    return item.name === 'Book';
});
console.log("Found Item 'Book':", foundItem);

// 4. REDUCE: Reduces the array to a single value
// Calculate total price of all items
// Syntax: array.reduce((accumulator, currentItem) => { ... }, initialValue)
const total = items.reduce((currentTotal, item) => {
    return item.price + currentTotal;
}, 0);
console.log("Total Price (Reduce):", total);

// 5. forEach: Executes a provided function once for each array element (No return value)
console.log("forEach Output:");
items.forEach((item) => {
    console.log(`- ${item.name}: $${item.price}`);
});
